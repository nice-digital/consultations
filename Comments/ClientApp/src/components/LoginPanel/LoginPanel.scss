@use "@nice-digital/nds-core/scss/typography";

.login-panel {
  &__back {
    border: none;
    cursor: pointer;
    display: inline;
    padding: 0;

    @include typography.links-default;
  }
}
