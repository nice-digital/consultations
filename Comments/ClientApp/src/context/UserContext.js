// @flow

import React from "react";
import Cookies from "js-cookie";
import { withRouter } from "react-router";

import { load } from "../data/loader";
import preload from "../data/pre-loader";

export const UserContext = React.createContext({
	updateContext: () => {},
});

type PropsType = {
	location: any,
	children: any,
	staticContext: any,
	match: any,
};

type StateType = {
	isAuthorised: boolean,
	isOrganisationCommenter: boolean,
	isLead: boolean,
	displayName: string,
	signInURL: string,
	registerURL: string,
	organisationName: string,
	initialDataLoaded: boolean,
};

export class UserProvider extends React.Component<PropsType, StateType> {
	constructor(props: PropsType) {
		super(props);

		this.state = {
			isAuthorised: false,
			isOrganisationCommenter: false,
			isLead: false,
			displayName: "",
			signInURL: "",
			registerURL: "",
			updateContext: this.updateContext,
			organisationName: null,
			initialDataLoaded: false,
		};

		const isServerSideRender = (this.props.staticContext && this.props.staticContext.preload);
		const preloadSource = isServerSideRender ? this.props.staticContext.preload.data : window.__PRELOADED__; // TODO: extract this preloaded line out to (or near) the preload endpoint method.

		const userSessionParameters = this.getUserSessionParameters(preloadSource);

		let isOrganisationCommenter = false;
		let organisationName = null;
		let isAuthorised = preloadSource ? preloadSource.isAuthorised : false;

		if (userSessionParameters.sessionCookieExistsForThisConsultation){
			const preloadedUserSessionData = preload(
				this.props.staticContext,
				"checkorganisationusersession",
				[],
				{
					consultationId: userSessionParameters.consultationId,
				},
				preloadSource,
			);

			if (preloadedUserSessionData){
				isOrganisationCommenter = preloadedUserSessionData.valid;
				isAuthorised = isAuthorised || preloadedUserSessionData.valid ; //you could be authorised by idam or by organisation session cookie.
				organisationName = preloadedUserSessionData.organisationName;
			}
		}

		if (preloadSource){
			this.state = {
				isAuthorised,
				isOrganisationCommenter: isOrganisationCommenter,
				isLead: preloadSource.isLead,
				displayName: preloadSource.displayName,
				signInURL: preloadSource.signInURL,
				registerURL: preloadSource.registerURL,
				updateContext: this.updateContext,
				organisationName: organisationName,
				initialDataLoaded: true,
			};
			if (this.props.staticContext) {
				this.props.staticContext.analyticsGlobals.isSignedIn = preloadSource.isAuthorised;
			}
		}
	}

	setStateForValidSessionCookie = (organisationName) => {
		this.setState({
			isAuthorised: true,
			isOrganisationCommenter: true,
			organisationName,
		});
	}

	loadUser = (returnURL) => {
		load("user", undefined, [], { returnURL, cachebust: new Date().getTime() })
			.then(
				res => {
					const signInURL = res.data.signInURL;
					this.setState({
						isAuthorised: res.data.isAuthenticatedByAccounts,
						displayName: res.data.displayName,
						signInURL: signInURL,
						registerURL: res.data.registerURL,
						isLead: (res.data.organisationsAssignedAsLead && res.data.organisationsAssignedAsLead.length > 0),
					});
					//update signin links in global nav here. because SSR isn't rendering them right on the server.
					var signInLinks = document.getElementById("global-nav-header").querySelectorAll("a[href*='account/login']");
					for (var i=0; i < signInLinks.length; i++) {
						signInLinks[i].setAttribute("href", signInURL);
					}
				},
			)
			.then(
				this.checkSessionId()
					.then(data => {
						if (data.validityAndOrganisationName?.valid === true) {
							this.setStateForValidSessionCookie(data.validityAndOrganisationName.organisationName);
						}
					}),
			);
	}

	getFeatureFlags = async () => {
		load("getfeatureflags", undefined, [], {})
			.then(response =>
				this.setState({
					organisationalCommentingFeature: response.data.OrganisationalCommenting,
				}))
			.catch(err => {
				console.log(JSON.stringify(err));
			});
	}

	//unfortunately the context is above the routes, so this.props.match is always null, so we can't pull the consultation id out of there. hence we're falling back to regex.
	getConsultationId = () =>{
		const regex = /^\/(\d+)\/(review|submitted|\d+\/[a-z-A-Z0-9]+)/;
		const pathname = this.props.location.pathname;

		if (!pathname)
			return;

		const matches = regex.exec(pathname);
		if (!matches || matches.length !== 3)
			return;

		return matches[1];
	}

	checkSessionId = async () => {
		const userSessionParameters = this.getUserSessionParameters(window.__PRELOADED__);

		if (!userSessionParameters.sessionCookieExistsForThisConsultation)
			return await {
				validityAndOrganisationName: {valid: false},
				userSessionParameters: userSessionParameters,
			};

		const validityAndOrganisationName = load(
			"checkorganisationusersession",
			undefined,
			[],
			{
				consultationId: userSessionParameters.consultationId,
			})
			.then(response => response.data)
			.catch(err => {
				console.log(JSON.stringify(err));
				return {valid: false};
			});
		return {
			validityAndOrganisationName: await validityAndOrganisationName,
			userSessionParameters: userSessionParameters,
		};
	}

	getUserSessionParameters = (preloadSource) => {
		const consultationId = this.getConsultationId();
		const cookieName = `ConsultationSession-${consultationId}`;
		const cookieString = preloadSource?.cookies || "";
		const sessionCookieExistsForThisConsultation =  (Cookies.get(cookieName) || cookieString.indexOf(cookieName) !== -1);
		return {consultationId, sessionCookieExistsForThisConsultation};
	}

	//a child component can call this method to update the context in case a session cookie has been set.
	updateContext = () => {
		this.checkSessionId()
			.then(data => {
				if (data.validityAndOrganisationName.valid === true) {
					this.setStateForValidSessionCookie(data.validityAndOrganisationName.organisationName);
				}
			});
	}

	loadContext = () => {
		this.loadUser(this.props.location.pathname);
	}

	componentDidMount() {
		this.loadContext(); //this is needed here as the sign in url isn't right on SSR. TODO: fix SSR.
	}

	componentDidUpdate(prevProps) {
		if (this.props.location !== prevProps.location) { //fired when the route changes, as you might be authenticated for some routes, but not others (e.g. org commenting cookie)
			this.loadContext();
		}
	}

	render() {
		return (
			<UserContext.Provider value={this.state}>
				{this.props.children}
			</UserContext.Provider>
		);
	}
}

export default withRouter(UserProvider);
