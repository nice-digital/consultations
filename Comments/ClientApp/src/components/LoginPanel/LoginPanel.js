// @flow
import React, { Component } from "react";
import { withRouter } from "react-router-dom";

import { UserContext } from "../../context/UserContext";
import { tagManager } from "../../helpers/tag-manager";

import  LoginBannerWithRouter from "../LoginBanner/LoginBanner";

type PropsType = {
	questionsTabIsOpen: boolean,
}

type StateType = {
	respondingAsOrg: boolean | null,
	respondingAsOrgType: string | null,
}

export class LoginPanel extends Component<PropsType, StateType> {

	constructor(props: PropsType) {
		super(props);

		this.state = {
			respondingAsOrg: null,
			respondingAsOrgType: null,
		};
	}

	componentDidMount() {}

	fieldsChangeHandler = (e) => {
		const fieldName = e.target.name;
		let fieldValue = e.target.value;

		const tagManagerLabel = {
			sole: "organisation login sole representative",
			lead: "organisation login as lead",
			code: "organisation login with code",
			false: "individual login",
			true: null,
		}[fieldValue];

		if (fieldValue === "true" || fieldValue === "false") {
			fieldValue = (fieldValue === "true");
		}

		this.setState(() => {
			if (tagManagerLabel) {
				tagManager({
					event: "generic",
					category: "Consultation comments page",
					action: "Clicked",
					label: tagManagerLabel,
				});
			}

			return { [fieldName]: fieldValue };
		});
	};

	goBack = () => {
		const { respondingAsOrgType } = this.state;

		// back to second question
		let newState = { respondingAsOrgType: null };

		// back to first question
		if (respondingAsOrgType === null || respondingAsOrgType === "code") {
			newState.respondingAsOrg = null;
		}

		this.setState(newState);
	};

	render() {
		const { respondingAsOrg, respondingAsOrgType } = this.state;
		const questionsTabIsOpen = this.props.questionsTabIsOpen;

		const showFirstScreen = (respondingAsOrg === null);
		const showSecondScreen = (respondingAsOrg === true && (respondingAsOrgType === null || respondingAsOrgType === "code"));
		const showLogin = (respondingAsOrg === false || respondingAsOrgType !== null);
		const showBackLink = (!showFirstScreen);

		const loginBannerProps = {
			signInButton: false,
			signInText: "to access the consultation",
			currentURL: this.props.match.url,
			signInURL: this.context.signInURL,
			registerURL: this.context.registerURL,
			allowOrganisationCodeLogin: false,
			title: "Make or review comments as an individual",
			isInCommentsPanel: true,
		};

		if (respondingAsOrgType === "lead" || respondingAsOrgType === "sole") {
			loginBannerProps.title = "Make or review comments on behalf of your organisation";
		}

		if (respondingAsOrgType === "code") {
			loginBannerProps.allowOrganisationCodeLogin = true;
			loginBannerProps.codeLoginOnly = true;
			loginBannerProps.title = null;
		}

		return (
			<>
				{showFirstScreen &&
					<LoginSelectionOrg
						currentlySelected={respondingAsOrg}
						onChangeOption={this.fieldsChangeHandler}
					/>
				}

				{showSecondScreen &&
					<LoginSelectionCode
						currentlySelected={respondingAsOrgType}
						onChangeOption={this.fieldsChangeHandler}
					/>
				}

				{showLogin &&
					<LoginBannerWithRouter {...loginBannerProps} />
				}

				{showBackLink &&
					<div className="container">
						<button className="login-panel__back" id="loginPanelBackButton" onClick={this.goBack}>&lt; Back</button>
					</div>
				}

				{questionsTabIsOpen &&
					<hr/>
				}
			</>
		);

	}
}

const LoginSelectionOrg = (props) => {
	const { currentlySelected, onChangeOption } = props;

	const radios = [
		{
			variant: "organisation",
			label: "As part of an organisation",
			value: true,
		},
		{
			variant: "individual",
			label: "As an individual",
			value: false,
		},
	];

	return (
		<div className="container" id="loginPanelScreen1">
			<h3>How are you taking part in this consultation?</h3>
			<p>You will not be able to change how you comment later.</p>
			<LoginSelectionRadio
				name="respondingAsOrg"
				radios={radios}
				currentlySelected={currentlySelected}
				onChangeOption={onChangeOption}
			/>
		</div>
	);
};

const LoginSelectionCode = (props) => {
	const { currentlySelected, onChangeOption } = props;

	const radios = [
		{
			variant: "code",
			label: "Yes, I have a code provided by my organisation",
			value: "code",
		},
		{
			variant: "lead",
			label: "Yes, I am the commenting lead for my organisation",
			value: "lead",
		},
		{
			variant: "sole",
			label: "No, I am the sole representative for my organisation",
			value: "sole",
		},
	];

	return (
		<div className="container" id="loginPanelScreen2">
			<h3>Do you have a code from your organisation?</h3>
			<LoginSelectionRadio
				name="respondingAsOrgType"
				radios={radios}
				currentlySelected={currentlySelected}
				onChangeOption={onChangeOption}
			/>
		</div>
	);
};

const LoginSelectionRadio = (props) => {
	const { name, radios, currentlySelected, onChangeOption } = props;

	return (
		<div role="radiogroup">
			{radios.map((radio, index) => (
				<div className="form__group form__group--radio form__group--inline" key={index}>
					<input
						className="form__radio"
						id={`${name}--${radio.variant}`}
						type="radio"
						name={name}
						checked={currentlySelected === radio.value}
						onChange={onChangeOption}
						value={radio.value}
					/>
					<label
						className="form__label form__label--radio"
						htmlFor={`${name}--${radio.variant}`}
						data-qa-sel={`loginpanel-label-${radio.variant}`}>
						{radio.label}
					</label>
				</div>
			))}
		</div>
	);
};

export default withRouter(LoginPanel);

LoginPanel.contextType = UserContext;
