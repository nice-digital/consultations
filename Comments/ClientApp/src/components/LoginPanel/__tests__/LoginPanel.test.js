import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import { UserContext } from "../../../context/UserContext";
import { LoginPanel } from "../LoginPanel";

const loginPanelFakeProps = {
	match: {
		url: "/1/1/introduction",
	},
};

test("shows first menu screen when first loaded", () => {
	render(
		<UserContext.Provider value={{}}>
			<LoginPanel {...loginPanelFakeProps} />
		</UserContext.Provider>,
	);
	expect(screen.getByRole("heading", { name: "How are you taking part in this consultation?" })).toBeInTheDocument();
});

it("shows second menu screen when organisation option has been selected", () => {
	render(
		<UserContext.Provider value={{}}>
			<LoginPanel { ...loginPanelFakeProps } />
		</UserContext.Provider>,
	);
	const radioButton = screen.getByLabelText("As part of an organisation");
	expect(screen.getByRole("heading", { name: "How are you taking part in this consultation?" })).toBeInTheDocument();
	fireEvent.click(radioButton);
	expect(screen.queryAllByRole("heading", { name: "How are you taking part in this consultation?" }).length).toBe(0);
	expect(screen.getByRole("heading", { name: "Do you have a code from your organisation?" })).toBeInTheDocument();
});

it("shows login when individual option has been selected", () => {
	render(
		<UserContext.Provider value={{}}>
			<MemoryRouter>
				<LoginPanel { ...loginPanelFakeProps } />
			</MemoryRouter>
		</UserContext.Provider>,
	);
	const radioButton = screen.getByLabelText("As an individual");
	expect(screen.getByRole("heading", { name: "How are you taking part in this consultation?" })).toBeInTheDocument();
	fireEvent.click(radioButton);
	expect(screen.queryAllByRole("heading", { name: "How are you taking part in this consultation?" }).length).toBe(0);
	expect(screen.getByRole("heading", { name: "Make or review comments as an individual" })).toBeInTheDocument();
});

it("shows second screen and organisation code text box when 'user has code' is selected", () => {
	render(
		<UserContext.Provider value={{}}>
			<MemoryRouter>
				<LoginPanel { ...loginPanelFakeProps } />
			</MemoryRouter>
		</UserContext.Provider>,
	);
	const radioButton = screen.getByLabelText("As part of an organisation");
	expect(screen.getByRole("heading", { name: "How are you taking part in this consultation?" })).toBeInTheDocument();
	fireEvent.click(radioButton);
	expect(screen.queryAllByRole("heading", { name: "How are you taking part in this consultation?" }).length).toBe(0);
	expect(screen.getByRole("heading", { name: "Do you have a code from your organisation?" })).toBeInTheDocument();
	const radioButton2 = screen.getByLabelText("Yes, I have a code provided by my organisation");
	fireEvent.click(radioButton2);
	expect(screen.getByLabelText("Enter your organisation code")).toBeInTheDocument();
});

it("goes back to previous screen when back has been clicked", () => {
	render(
		<UserContext.Provider value={{}}>
			<MemoryRouter>
				<LoginPanel { ...loginPanelFakeProps } />
			</MemoryRouter>
		</UserContext.Provider>,
	);
	const radioButton = screen.getByLabelText("As an individual");
	expect(screen.getByRole("heading", { name: "How are you taking part in this consultation?" })).toBeInTheDocument();
	fireEvent.click(radioButton);
	expect(screen.queryAllByRole("heading", { name: "How are you taking part in this consultation?" }).length).toBe(0);
	expect(screen.getByRole("heading", { name: "Make or review comments as an individual" })).toBeInTheDocument();
	const backButton = screen.getByText(/Back/i, { selector: "button" });
	fireEvent.click(backButton);
	expect(screen.queryAllByRole("heading", { name: "Make or review comments as an individual" }).length).toBe(0);
	expect(screen.getByRole("heading", { name: "How are you taking part in this consultation?" })).toBeInTheDocument();
});

it("clears radio buttons after navigating back to previous screen", () => {
	render(
		<UserContext.Provider value={{}}>
			<MemoryRouter>
				<LoginPanel { ...loginPanelFakeProps } />
			</MemoryRouter>
		</UserContext.Provider>,
	);
	const radioButton = screen.getByLabelText("As an individual");
	expect(screen.getByRole("heading", { name: "How are you taking part in this consultation?" })).toBeInTheDocument();
	fireEvent.click(radioButton);
	expect(screen.queryAllByRole("heading", { name: "How are you taking part in this consultation?" }).length).toBe(0);
	expect(screen.getByRole("heading", { name: "Make or review comments as an individual" })).toBeInTheDocument();
	const backButton = screen.getByText(/Back/i, { selector: "button" });
	fireEvent.click(backButton);
	expect(screen.queryAllByRole("heading", { name: "Make or review comments as an individual" }).length).toBe(0);
	expect(screen.getByRole("heading", { name: "How are you taking part in this consultation?" })).toBeInTheDocument();
	const allRadioButtons = screen.queryAllByRole("radio");
	allRadioButtons.map(node => {
		expect(node.checked).toEqual(false);
	});
});

it("should match the snapshot after first loading up", () => {
	const {container} = render(
		<UserContext.Provider value={{}}>
			<MemoryRouter>
				<LoginPanel { ...loginPanelFakeProps } />
			</MemoryRouter>
		</UserContext.Provider>,
	);
	expect(container).toMatchSnapshot();
});
