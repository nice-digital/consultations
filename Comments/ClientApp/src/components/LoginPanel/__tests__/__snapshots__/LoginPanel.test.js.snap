// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot after first loading up 1`] = `
<div>
  <div
    class="container"
    id="loginPanelScreen1"
  >
    <h3>
      How are you taking part in this consultation?
    </h3>
    <p>
      You will not be able to change how you comment later.
    </p>
    <div
      role="radiogroup"
    >
      <div
        class="form__group form__group--radio form__group--inline"
      >
        <input
          class="form__radio"
          id="respondingAsOrg--organisation"
          name="respondingAsOrg"
          type="radio"
          value="true"
        />
        <label
          class="form__label form__label--radio"
          data-qa-sel="loginpanel-label-organisation"
          for="respondingAsOrg--organisation"
        >
          As part of an organisation
        </label>
      </div>
      <div
        class="form__group form__group--radio form__group--inline"
      >
        <input
          class="form__radio"
          id="respondingAsOrg--individual"
          name="respondingAsOrg"
          type="radio"
          value="false"
        />
        <label
          class="form__label form__label--radio"
          data-qa-sel="loginpanel-label-individual"
          for="respondingAsOrg--individual"
        >
          As an individual
        </label>
      </div>
    </div>
  </div>
</div>
`;
